src/api/eth_filter.rs

```diff
@@ -1,10 +1,403 @@
+    EthFilter {
+      transport: transport,
+    }
+  }
+}
+
+impl<T: Transport + Clone> EthFilter<T> {
+  /// Installs a new logs filter.
+  pub fn new_logs_filter(&self, filter: Filter) -> FilterResult<T, CallResult<U256, T::Out>, LogsFilter<T>> {
+    let f = helpers::serialize(&filter);
+    FilterResult::new(
+      self.transport.clone(),
+      CallResult::new(self.transport.execute("eth_newFilter", vec![f]))
+    )
+  }
+
+  /// Installs a new block filter.
+  pub fn new_blocks_filter(&self) -> FilterResult<T, CallResult<U256, T::Out>, BlocksFilter<T>> {
+    FilterResult::new(
+      self.transport.clone(),
+      CallResult::new(self.transport.execute("eth_newFilter", vec![]))
+    )
+  }
+
+  /// Installs a new pending transactions filter.
+  pub fn new_pending_transactions_filter(&self) -> FilterResult<T, CallResult<U256, T::Out>, PendingTransactionsFilter<T>> {
+    FilterResult::new(
+      self.transport.clone(),
+      CallResult::new(self.transport.execute("eth_newFilter", vec![]))
+    )
+  }
+}
+
+/// Result of installing new filter
+pub struct FilterResult<T, F, X> {
+  transport: T,
+  inner: F,
+  _output: PhantomData<X>,
+}
+
+impl<T, F, X> FilterResult<T, F, X> {
+  /// Create a new CallResult wrapping the inner future.
+  pub fn new(transport: T, inner: F) -> Self {
+    FilterResult { transport: transport, inner: inner, _output: PhantomData }
+  }
+}
+
+impl<T, F, X> Future for FilterResult<T, F, X> where
+  T: Transport + Clone,
+  F: Future<Item=U256, Error=Error>,
+  X: FilterHandle<T>,
+{
+  type Item = X;
+  type Error = Error;
+
+  fn poll(&mut self) -> Poll<X, Error> {
+    match self.inner.poll() {
+      Ok(Async::Ready(x)) => Ok(Async::Ready(X::new(x, self.transport.clone()))),
+      Ok(Async::NotReady) => Ok(Async::NotReady),
+      Err(e) => Err(e),
+    }
+  }
+}
+
+/// Generic filter interface
+pub trait FilterHandle<T: Transport> {
+  /// Creates a new filter.
+  fn new(id: U256, transport: T) -> Self;
+}
+
+/// Base filter handle.
+/// Uninstall filter on drop.
+/// Allows to poll the filter.
+pub struct BaseFilter<T: Transport, R> {
+  id: U256,
+  transport: T,
+  _output: PhantomData<R>,
+}
+
+impl<T: Transport, R> BaseFilter<T, R> {
+  /// Polls this filter for changes.
+  /// Will return logs that happened after previous poll.
+  pub fn poll(&self) -> CallResult<Option<Vec<R>>, T::Out> {
+    let id = helpers::serialize(&self.id);
+    CallResult::new(self.transport.execute("eth_getFilterChanges", vec![id]))
+  }
+
+  /// Uninstalls the filter
+  pub fn uninstall(self) -> CallResult<bool, T::Out> where Self: Sized {
+    self.uninstall_internal()
+  }
+
+  fn uninstall_internal(&self) -> CallResult<bool, T::Out> {
+    let id = helpers::serialize(&self.id);
+    CallResult::new(self.transport.execute("eth_uninstallFilter", vec![id]))
+  }
+}
+
+impl<T: Transport, R> Drop for BaseFilter<T, R> {
+  fn drop(&mut self) {
+    let _ = self.uninstall_internal().wait();
```

Comment:
I think it needs to be polled couple of times (you cannot tell how many, cause it might depend on transport and io availability)